cli/internal/codegen/testdata/expected_typescript.ts

```diff
@@ -385,34 +399,324 @@ class BaseClient {
         // Merge our headers with any predefined headers
         init.headers = {...this.headers, ...init.headers}
 
+        let bearerToken: string | undefined
+
         // If an authorization token generator is present, call it and add the returned token to the request
         if (this.tokenGenerator) {
-            init.headers["Authorization"] = "Bearer " + this.tokenGenerator()
+            bearerToken = this.tokenGenerator()
+        }
+
+        // If we now have a bearer token, add it to the request
+        if (bearerToken) {
+            init.headers["Authorization"] = "Bearer " + bearerToken
+        } else if (requiresAuth) {
+            // The API called requires authorization, but we don't have a token generator, or it provided a blank token value
+            throw new APIError(400, { code: ErrCode.Unauthenticated, message: 'Authorization token required for this API, but none provided.' })
         }
 
         // Make the actual request
         const response = await this.fetcher(this.baseURL + path, init)
 
         // handle any error responses
         if (!response.ok) {
-            const body = await response.text()
-            throw new Error(`request failed: status ${response.status}: ${body}`)
+            // try and get the error message from the response body
+            let body: APIErrorResponse = { code: ErrCode.Unknown, message: `request failed: status ${response.status}` }
+
+            // if we can get the structured error we should, otherwise give a best effort
+            try {
+                let jsonBody = await response.json()
+                if (isAPIErrorResponse(jsonBody)) {
+                    body = jsonBody
+                } else {
+                    body.message += ": " + JSON.stringify(jsonBody)
+                }
+            } catch {
+                // otherwise we just append the text to the error message
+                body.message += ": " + await response.text()
```

Comment:
If `response.json()` fails, I don't believe we can call `response.text()` as the body has already been consumed. Maybe better to call `response.text()` and then `JSON.parse()` on that?